src/core/qgsmaprenderertask.cpp

```diff
@@ -21,30 +21,114 @@
 #include "qgsmapsettingsutils.h"
 #include "qgsogrutils.h"
 #include "qgslogger.h"
+#include "qgsabstractgeopdfexporter.h"
+#include "qgsmaprendererstagedrenderjob.h"
+#include "qgsrenderedfeaturehandlerinterface.h"
+#include "qgsfeaturerequest.h"
+#include "qgsvectorlayer.h"
 #include <QFile>
 #include <QTextStream>
 #include <QPrinter>
 
 #include "gdal.h"
 #include "cpl_conv.h"
 
-QgsMapRendererTask::QgsMapRendererTask( const QgsMapSettings &ms, const QString &fileName, const QString &fileFormat, const bool forceRaster )
+///@cond PRIVATE
+
+class QgsMapRendererTaskGeoPdfExporter : public QgsAbstractGeoPdfExporter
+{
+
+  public:
+
+    QgsMapRendererTaskGeoPdfExporter( const QgsMapSettings &ms )
+    {
+      // collect details upfront, while we are still in the main thread
+      const QList< QgsMapLayer * > layers = ms.layers();
+      for ( const QgsMapLayer *layer : layers )
+      {
+        VectorComponentDetail detail;
+        detail.name = layer->name();
+        detail.mapLayerId = layer->id();
+        if ( const QgsVectorLayer *vl = qobject_cast< const QgsVectorLayer * >( layer ) )
+        {
+          detail.displayAttribute = vl->displayField();
+        }
+        mLayerDetails[ layer->id() ] = detail;
+      }
+    }
+
+  private:
+
+    QgsAbstractGeoPdfExporter::VectorComponentDetail componentDetailForLayerId( const QString &layerId ) override
+    {
+      return mLayerDetails.value( layerId );
+    }
+
+    QMap< QString, VectorComponentDetail > mLayerDetails;
+};
+
+
+class QgsMapRendererTaskRenderedFeatureHandler : public QgsRenderedFeatureHandlerInterface
+{
+  public:
+
+    QgsMapRendererTaskRenderedFeatureHandler( QgsMapRendererTaskGeoPdfExporter *exporter )
+      : mExporter( exporter )
+    {}
+
+    void handleRenderedFeature( const QgsFeature &feature, const QgsGeometry &renderedBounds, const QgsRenderedFeatureHandlerInterface::RenderedFeatureContext &context ) override
+    {
+      // is it a hack retrieving the layer ID from an expression context like this? possibly... BUT
+      // the alternative is adding a layer ID member to QgsRenderContext, and that's just asking for people to abuse it
+      // and use it to retrieve QgsMapLayers mid-way through a render operation. Lesser of two evils it is!
+      const QString layerId = context.renderContext.expressionContext().variable( QStringLiteral( "layer_id" ) ).toString();
+
+      // transform from pixels to PDF coordinates (TODO - check)
+      QTransform pixelToPdfTransform = QTransform::fromScale( 25.4 / context.renderContext.scaleFactor() / 72.0, 25.4 / context.renderContext.scaleFactor() / 72.0 );
```

Comment:
This is assuming a DPI of 72, right? Is it the case that 'PDF coordinates' are hardcoded to 72 DPI?

Reply:
Yes, added a comment